pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_sni.go

```diff
@@ -0,0 +1,256 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubeapiserverexposure
+
+import (
+	"bytes"
+	"context"
+	_ "embed"
+	"fmt"
+	"text/template"
+	"time"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/operation/botanist/component"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/vpnseedserver"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/Masterminds/sprig"
+	protobuftypes "github.com/gogo/protobuf/types"
+	istioapinetworkingv1beta1 "istio.io/api/networking/v1beta1"
+	istionetworkingv1alpha3 "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	istionetworkingv1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// SNIValues configure the kube-apiserver service SNI.
+type SNIValues struct {
+	Hosts                    []string
+	NamespaceUID             types.UID
+	APIServerClusterIP       string
+	APIServerInternalDNSName string
+	IstioIngressGateway      IstioIngressGateway
+	ReversedVPN              ReversedVPN
+}
+
+// IstioIngressGateway contains the values for istio ingress gateway configuration.
+type IstioIngressGateway struct {
+	Namespace string
+	Labels    map[string]string
+}
+
+// ReversedVPN contains whether the reversed vpn is enabled or not.
+type ReversedVPN struct {
+	Enabled bool
+}
+
+// NewSNI creates a new instance of DeployWaiter which deploys Istio resources for
+// kube-apiserver SNI access.
+func NewSNI(
+	client client.Client,
+	applier kubernetes.Applier,
+	namespace string,
+	values *SNIValues,
+) component.DeployWaiter {
+	if values == nil {
+		values = &SNIValues{}
+	}
+
+	return &sni{
+		client:    client,
+		applier:   applier,
+		namespace: namespace,
+		values:    values,
+	}
+}
+
+type sni struct {
+	client    client.Client
+	applier   kubernetes.Applier
+	namespace string
+	values    *SNIValues
+}
+
+func (s *sni) Deploy(ctx context.Context) error {
+	var (
+		destinationRule = s.emptyDestinationRule()
+		envoyFilter     = s.emptyEnvoyFilter()
+		gateway         = s.emptyGateway()
+		virtualService  = s.emptyVirtualService()
+
+		hostName                  = fmt.Sprintf("%s.%s.svc.%s", v1beta1constants.DeploymentNameKubeAPIServer, s.namespace, gardencorev1beta1.DefaultDomain)
+		envoyFilterSpec           bytes.Buffer
+		envoyFilterTemplateValues = map[string]interface{}{
```

Comment:
Can you add struct for the values instead of free-style `map[string]interface`?

Reply:
done with c128c00